flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherTest.java

```diff
@@ -1045,6 +1048,27 @@ public void testJobDataAreCleanedUpInCorrectOrderOnFailedJob() throws Exception
         testJobDataAreCleanedUpInCorrectOrder(JobStatus.FAILED);
     }
 
+    @Test
+    public void testOnlyRecoveredJobsAreRetainedInTheBlobServer() throws Exception {
+        final JobID jobId1 = new JobID();
+        final JobID jobId2 = new JobID();
+        final byte[] fileContent = {1, 2, 3, 4};
+        final BlobServer blobServer = getBlobServer();
+        final PermanentBlobKey blobKey1 = blobServer.putPermanent(jobId1, fileContent);
+        final PermanentBlobKey blobKey2 = blobServer.putPermanent(jobId2, fileContent);
+
+        final TestingDispatcher dispatcher =
+                new TestingDispatcherBuilder()
+                        .setInitialJobGraphs(Collections.singleton(new JobGraph(jobId1, "foobar")))
+                        .build();
+
+        org.assertj.core.api.Assertions.assertThat(blobServer.getFile(jobId1, blobKey1))
```

Comment:
This should _should_ now be unstable since the retain runs in a separate executor.

Reply:
True. I will use a direct executor to solve the problem.